### PR TITLE
renameFile now consistently reports an error if the destination is a directory

### PR DESCRIPTION
**Pull request by gintas**
I only tested on Windows, could you please run the tests on Unix to make sure they are not broken?

https://ghc.haskell.org/trac/ghc/ticket/8482 is related.

### CODE REVIEW

**Comment by dcoutts** [comment]
This looks like a similar issue as we're discussing in #10 and #4, that after an operation we need to check if it's a dir or not to distinguish certain error cases. The subtlety is what to do when that "is it a dir" check itself fails (e.g. because it does not exist, or has a perm error or whatever). We should look around for other instances of this within `directory` and try and deal with them consistently.

File: System/Directory.hs [hunk 1]
```diff
@@ -638,24 +638,31 @@ Either path refers to an existing directory.
...
-   if is_dir
-        then ioError (ioeSetErrorString
-                          (mkIOError InappropriateType "renameFile" Nothing (Just opath))
-                          "is a directory")
-        else do
+   -- XXX the isDirectory tests are not performed atomically with the rename
+   checkNotDir opath
+   doRename `E.catch` renameExcHandler
+ where checkNotDir path = do
+                isdir <- pathIsDir path `E.catch` ((\ _ -> return False) :: IOException -> IO Bool)
```
Comment (Rufflewind) [inline System/Directory.hs]: You can use `catchIOError` to avoid specifying the type signature.

File: System/Directory.hs [hunk 2]
```diff
@@ -638,24 +638,31 @@ Either path refers to an existing directory.
...
-   let is_dir = Posix.isDirectory stat
-#endif
-   if is_dir
-        then ioError (ioeSetErrorString
-                          (mkIOError InappropriateType "renameFile" Nothing (Just opath))
-                          "is a directory")
-        else do
+   -- XXX the isDirectory tests are not performed atomically with the rename
+   checkNotDir opath
+   doRename `E.catch` renameExcHandler
```
Comment (Rufflewind) [inline System/Directory.hs]: I think `renameExcHandler` should be omitted for non-Windows systems.

**Comment by Rufflewind** [comment]
Certain race conditions can still cause bug to reappear, but I'm not sure if there's a way to avoid this problem entirely.

**Pull request merged**